### ISSUE TITLE
Conda-based Installer Parent Issue

### ISSUE THREAD

**Issue by mrclary** [spyder-ide/spyder#20001]
This issue is intended to capture discussions related to the development of the conda-based standalone Spyder installers post release of Spyder 5.4.0.

This description will aim to capture the up-to-date status of the development strategy resulting from a consensus of comments. Bug fixes and explicit tasks should be reported in a separate issue and PR.

# Outstanding Questions, Ideas, & Proposals
1. Update the Spyder conda installation as follows:
   - Make the root conda environment the Spyder runtime environment
   - Name the root conda environment directory `spyder-X.X` with major and minor version only
   - The root environment will support all bug fix releases within the minor release, and can be updated via conda-forge
   - New minor and major releases will require downloading a new installer.
1. Update the Spyder conda installation as follows:
   - Make the root conda environment the Spyder runtime environment
   - Name the root conda environment directory `spyder-X` with major version only
   - The root environment will support all bug fix and feature releases, and can be updated via conda-forge
   - New major releases will require downloading a new installer.
1. All updates, including those requiring new installer, can be managed by Spyder's update manager
# To Implement
* [x] Linux installer suggestions, see [Suggestions from @ccordoba12](https://github.com/spyder-ide/spyder/issues/20176#issuecomment-1366006777) and #20319
   * Add command line alias to launch Spyder
   * Create uninstall script and alias for command line
   * Add message at the end of installation providing launch instructions
   * Update the application icon
* [x] Add installer tests for Linux and Windows #20601 
* [x] Remove options from Windows installer #20838 :
   * Add Spyder to my PATH environment variable
   * Register Spyder as my default Python 3.10
* [x] Group taskbar icons on Windows #20791 conda/menuinst#127
* [x] Prevent showing a cmd window when launching Spyder from the installer shorcut conda/menuinst#128
* [x] Installer should ensure that Spyder is closed. Cannot be done with `constructor`. Implemented with `UpdateManager` plugin (#21483)
* [x] Installer should provide option to uninstall existing conflicting installation. Cannot be done with `constructor`. Implemented with `UpdateManager` plugin (#21483)
* [x] #21640
* [x] #21647
* [x] #21782
* [x] #21785
* [x] #21836
* [x] #21871
* [x] #22059
* [x] #22072
* [ ] Open files from Windows Explorer (maybe the new registry handling available from `menuinst` could help: https://github.com/conda/menuinst/pull/117 there is a TODO there about context menu entries). Also conda/menuinst#185 and conda/menuinst#179. conda-forge/spyder-feedstock#171

**Comment by ccordoba12** [comment]
I have one comment about this: I think the strategy of using git shallow clones in our subrepo conda recipes doesn't work when developers sync the subrepos to a branch in their forks (I saw that in the CIs of PR #19992 yesterday although is not visible anymore because of a rebase).

That's because developers usually don't care to push tags to their forks. So I'd suggest to remove shallow clones and clone subrepos in full instead.

**Comment by mrclary** [comment]
> That's because developers usually don't care to push tags to their forks. So I'd suggest to remove shallow clones and clone subrepos in full instead.

Good point. I'll make the change.

**Comment by ccordoba12** [comment]
Another problem appeared: The Spyder-kernels patch applied to the recipe used for the installers no longer applies. This is due to https://github.com/conda-forge/spyder-kernels-feedstock/pull/86.

**Comment by mrclary** [comment]
> Another problem appeared: The Spyder-kernels patch applied to the recipe used for the installers no longer applies. This is due to [conda-forge/spyder-kernels-feedstock#86](https://github.com/conda-forge/spyder-kernels-feedstock/pull/86).

Thanks for giving me the heads up. I'll update the patch for building spyder-kernels. Also, since it is now noarch, it can be built in parallel with the other noarch dependencies; I'll update the workflow to accommodate.

**Comment by ccordoba12** [comment]
Another suggestion: Could we only build the Intel app in PRs for Mac? That should be enough, given that it can run on M1/M2 chips, right?

The thing is we only have 5 Mac slots for the entire organization, and right now we're using more than that on every PR here, which slows down development in other projects.

**Comment by mrclary** [comment]
> Another suggestion: Could we only build the Intel app in PRs for Mac? That should be enough, given that it can run on M1/M2 chips, right?
> 
> The thing is we only have 5 Mac slots for the entire organization, and right now we're using more than that on every PR here, which slows down development in other projects.

Sure thing. I'll remove arm64 from the build matrix.

**Comment by dalthviz** [comment]
Some notes regarding missing things or behaviors after reviewing the conda-based installer (for Windows) generated by https://github.com/spyder-ide/spyder/pull/20368 that probably we will need to check:

* Add a way to open files from the Windows Explorer:

![edit](https://user-images.githubusercontent.com/16781833/217660195-bde6ce09-20c6-45a2-9408-3bc7bed264e4.png)

* Prevent showing a cmd window when launching Spyder from the installer shorcut:

![image](https://user-images.githubusercontent.com/16781833/217662630-0ff5f31b-ed60-45ce-8e14-9fe2eb316f45.png)

* Add a way to prompt for an uninstall if the installer is run and Spyder is detected as installed (currently it shows a warning but only if the installer selected installation path is the same as the old installation):

![image](https://user-images.githubusercontent.com/16781833/217662767-3a4cfd3e-a9a2-45d0-88b8-11fb69f4ff9c.png)

* Not totally sure if some of the options the installer shows should be available (like adding Spyder to the PATH):

![image](https://user-images.githubusercontent.com/16781833/217663798-7b0c0a74-fcc5-4f72-b8b3-3a6fda6225e6.png)

* Add a way to run Spyder after the installation finishes:

![image](https://user-images.githubusercontent.com/16781833/217661816-760d0775-c27b-4fc4-97c4-96437ed4216d.png)

* Seems like there is an issue to correctly show the Python version of the default interpreter in the status bar (it shows `custom ()` sometimes): 

![image](https://user-images.githubusercontent.com/16781833/217662082-deb5e812-d2ad-4766-9fb4-54be2c7d9970.png)

* The installer messages and the installer name suggest the installer comes with Python 3.10, but checking the about dialog inside Spyder seems like is Python 3.9:

![image](https://user-images.githubusercontent.com/16781833/217665405-d622277a-b6d9-4f85-bba5-94bdc8da8856.png)

Also, probably we need to check the way the installation method is marked so the conda-based standalone installers are not marked as `conda` but `conda-based` or `standalone conda`.

Maybe most of these things can be done with pre/post install/uninstall scripts and/or menuinst @mrclary ? Also, is there maybe is there a way to pass to `constructor` a template for NSIS to be able to reuse the logic of the old installer?

**Comment by mrclary** [comment]
@dalthviz, thanks for all the observations. I don't see any issues implementing your suggestions.
Regarding the old installer logic, I'm not sure what that is. Perhaps we can review it at the developer meeting and I can see if it can be incorporated into the constructor; most likely it would be handled by pre/post-install scripts.

**Comment by dalthviz** [comment]
> Regarding the old installer logic, I'm not sure what that is

Is the logic to detect if Spyder is already installed and if it is running, prompt the user to close Spyder, or run the uninstaller and so on (which was done inside a NSIS template with the package that the old installer for Windows used to be build - `pynsist`).

Anyhow, I guess the way constructor customization works is by the pre/post scripts in general so we will need to figure out how that behavior can be ported to the Windows conda-based installer or even if is worthy to keep those validations or not.

**Comment by mrclary** [comment]
@dalthviz, I'm pretty sure that we can use an nsis template. If you can provide one, I'll try to implement it.

**Comment by ccordoba12** [comment]
@mrclary, some feedback about the Linux installer after testing the 5.4.3 one:

* I think it shouldn't offer the option to activate the installer environment in the console (that's the default right now).
* The `spyder` and `uninstall-spyder` aliases are not working for me in the console after installation (I don't know if that has to do with the fact that I selected the option to activate the env I mentioned above).

**Comment by ccordoba12** [comment]
Also, I'd like to ask you to take care of porting the update mechanism that @jsbautista and @dalthviz developed last year to the new installers. We need to make sure that that work as it is now before releasing Spyder 6 and @dalthviz is quite busy with migrating the Editor to the new API.

**Comment by mrclary** [comment]
> * The `spyder` and `uninstall-spyder` aliases are not working for me in the console after installation (I don't know if that has to do with the fact that I selected the option to activate the env I mentioned above).

If you open a new terminal tab do the aliases work? Or quit and reopen the terminal application?

**Comment by mrclary** [comment]
> @mrclary, some feedback about the Linux installer after testing the 5.4.3 one:
> 
> * I think it shouldn't offer the option to activate the installer environment in the console (that's the default right now).
> * The `spyder` and `uninstall-spyder` aliases are not working for me in the console after installation (I don't know if that has to do with the fact that I selected the option to activate the env I mentioned above).

I'll see if I can fix these in #20827

**Comment by ccordoba12** [comment]
> If you open a new terminal tab do the aliases work? Or quit and reopen the terminal application?

Nop, they don't.

**Comment by mrclary** [comment]
> I installed into a fresh `mamba` installation

@joooeey, can you clarify what you mean here? This new conda-based installer should be [downloaded](https://github.com/spyder-ide/spyder/releases/download/v5.4.3/EXPERIMENTAL-Spyder-5.4.3-Linux-x86_64.sh) and run
```
$ sh ~/Downloads/EXPERIMENTAL-Spyder-5.4.3-Linux-x86_64.sh
```
You should find a conda environment installed at `~/.local/Spyder-5.4.3`. Creating your own environment and installing Spyder via `$ mamba install spyder` will not produce application shortcuts.

**Comment by joooeey** [comment]
I used `$ mamba install spyder`

I'm not interested in downloading experimental versions and I would generally like to use my package manager (`mamba` because of its dependency resolution speed) for everything. If that doesn't produce a start menu shortcut, so be it, but it's a shame.

**Comment by ccordoba12** [comment]
@joooeey, these improvements are only available for our own Linux installer, not for the Spyder conda package, sorry.

**Comment by mrclary** [comment]
> I used `$ mamba install spyder`
> 
> I'm not interested in downloading experimental versions and I would generally like to use my package manager (`mamba` because of its dependency resolution speed) for everything. If that doesn't produce a start menu shortcut, so be it, but it's a shame.

@joooeey, just to clarify what our "experimental" installer does:
* It is not an experimental version of Spyder, just the _installer_ is experimental.
* The installer creates an independent conda (with mamba) environment with Spyder installed. This does not replace any existing conda or mamba installation you may have on your system.
* An application shortcut is created; ~/.local/share/applications/spyder-spyder.desktop for Linux, ~/Applications/Spyder.app for macOS.
* Any Python environment, including other conda environments, can still be used for Console environments, just like your traditionally installed Spyder.

**Comment by ccordoba12** [comment]
@mrclary, a couple of comments after seeing the results of the alpha1 release:

* We should use Python 3.10 for the installers because it's more performant than 3.9.
* The arm64 Mac installer was not created by the release workflow.

**Comment by mrclary** [comment]
Closing per 6.0.0 release
